Add events to schedule page

rileysw opened the pull request:
* Added events to the schedule page

TODOs
* Add countdown
* Add Bootstrap Scrollspy (for navigation by day)
* Add message at the top saying to refer to Slack announcements for locations

rileysw committed:
added events to schedule page

rileysw committed:
added key to mapped components

Pegysus committed:
added Countdown component to banner. It is a bit jank right now but I will try to fix css/sizing later if I have time (if anyone has time to help fix that'll be greatly appreciated)

Pegysus committed:
Update src/views/schedule/components/Countdown.tsx

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Pegysus committed:
Update src/views/schedule/Schedule.module.scss

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

taesungh commented on src/assets/data/schedule.json:
```diff
@@ -0,0 +1,368 @@
+		{
+			"title": "Check In",
+			"time": {
+				"start": "2023-02-03 17:00:00-08",
+				"end": "2023-02-03 18:00:00-08"
```
Issue: This date time format is not recognized by all browsers, notably Safari.

taesungh commented on src/views/schedule/Schedule.module.scss:
```diff
@@ -0,0 +1,11 @@
+@use "utils/bootstrap";
+@import "styles/gallery-theme";
```
Chore: stylesheet imports are unused

taesungh commented on src/views/schedule/Schedule.tsx:
```diff
@@ -0,0 +1,135 @@
+						})}
+						end={new Date(event.time.end).toLocaleTimeString([], {
+							hour: "2-digit",
+							minute: "2-digit",
+						})}
+						category={event.category}
+						host={event.host}
+						description={event.description}
+					/>
+				)
+			)}
+			{schedule.sunday.map((event) =>
+				event.category === "Announcement" ? (
+					<Announcement
+						key={event.title}
+						title={event.title}
+						start={new Date(event.time.start).toLocaleTimeString([], {
+							hour: "2-digit",
+							minute: "2-digit",
+						})}
+						description={event.description}
+					/>
+				) : (
+					<Event
+						key={event.title}
+						title={event.title}
+						start={new Date(event.time.start).toLocaleTimeString([], {
+							hour: "2-digit",
+							minute: "2-digit",
+						})}
+						end={new Date(event.time.end).toLocaleTimeString([], {
+							hour: "2-digit",
+							minute: "2-digit",
+						})}
+						category={event.category}
+						host={event.host}
+						description={event.description}
+					/>
+				)
+			)}
```
It seems we are repeating ourselves quite a lot here. If the schedule for each day would look the same, we could refactor this repeated logic.

taesungh commented on src/views/schedule/components/Announcement.tsx:
```diff
@@ -0,0 +1,29 @@
+	return (
+		<Card className={styles.card}>
+			<Card.Body>
+				<Card.Title as="h4">
+					{description !== "" ? (
+						<a className={styles.announcementLink} href={description}>
+							<h4>{title}</h4>
+						</a>
+					) : (
+						<h4>{title}</h4>
+					)}
+				</Card.Title>
```
Issue: Invalid DOM nesting
> `<h4>` cannot appear as a child of `<h4>`

taesungh commented on src/views/schedule/components/Event.tsx:
```diff
@@ -0,0 +1,33 @@
+	return (
+		<Card className={styles.card}>
+			<Card.Body>
+				<Card.Title as="h4" className={styles.cardTitle}>
+					<h4 className={styles.cardTitleName}>{title}</h4>
+					<h4 className={styles.cardTitleCategory}>{category}</h4>
+				</Card.Title>
```
Issue: Invalid DOM nesting
> `<h4>` cannot appear as a child of `<h4>`

taesungh commented on src/views/schedule/components/Event.module.scss:
```diff
@@ -0,0 +1,36 @@
+	color: $orange;
+}
+
+.cardTitleCategory {
+	display: flex;
+	text-align: center;
+	align-items: center;
+	justify-content: center;
+	border-radius: 30px;
+	padding: 0px 20px;
+	background-color: $light-orange;
+	color: $midnight-blue;
+}
```
Note: Bootstrap has a [Badge](https://getbootstrap.com/docs/5.3/components/badge/) component we can use

taesungh commented on src/views/schedule/components/Event.tsx:
```diff
@@ -0,0 +1,33 @@
+
+function Event({ title, start, end, category, host, description }: EventProps) {
+	return (
+		<Card className={styles.card}>
+			<Card.Body>
+				<Card.Title as="h4" className={styles.cardTitle}>
+					<h4 className={styles.cardTitleName}>{title}</h4>
+					<h4 className={styles.cardTitleCategory}>{category}</h4>
+				</Card.Title>
+				<Card.Subtitle className={styles.cardSubtitle}>
+					{host !== "" ? <p>Hosted by: {host}</p> : null}
+					<p>
+						{start} - {end}
+					</p>
+				</Card.Subtitle>
+				<Card.Text>{description}</Card.Text>
+			</Card.Body>
+		</Card>
```
Praise: thank you for making this using the Card component

taesungh commented on src/views/schedule/components/Countdown.tsx:
```diff
@@ -0,0 +1,120 @@
+		}
+	};
+
+	return countdownGenerator();
```
Suggestion: this component could just return what it needs to rather than using this nested function

taesungh commented on src/views/schedule/Schedule.tsx:
```diff
@@ -0,0 +1,135 @@
+	return (
+		<Container>
+			<TitleBanner>
+				{generateCountdown()}
```
Note: directly invoking functional components may lead to unexpected behavior

taesungh reviewed (changes requested):
Decent start. There are some rendering issues to fix and some reorganization that could be done.

taesungh committed:
Use `const` instead of `let` in `Countdown.tsx`

taesungh committed:
Refactor Schedule List and reduce repetition

taesungh committed:
Convert schedule event times to ISO format

- The previous format was not supported in Safari
- Date parsing is implementation dependent, so follow the ECMA262 spec
  which is a subset of ISO 8601

taesungh committed:
Use `Badge` component for event category labels

- Make categories lowercase
- Fix DOM nesting issue
- Rename `Event` to `EventCard`

taesungh committed:
Include relative schedule event times with Day.js

- Similar to HackUCI 2021 schedule, add schedule event card footer with
  relative time to/from event

rileysw committed:
changed hacking begins + submission times to PST

rileysw committed:
clean up code + eslint errors

rileysw committed:
fix invalid DOM nesting

rileysw committed:
changed hacking periods to events from announcements

rileysw committed:
add schedule to navbar

taesungh committed:
Create EventSpacerCard for long hacking periods

- Similar to 2021 schedule, Use tall spacer card for hacking periods
- Reorganize schedule item construction by spreading props

taesungh committed:
Fix props in schedule Announcement Card

taesungh committed:
Make past event cards transparent

- Similar to 2021 schedule, make cards for events that ended transparent
- Periodically refresh current time to update `now` reference time

taesungh committed:
Use tall title banner in schedule landing

- Make schedule landing similar to Home landing for schedule event cards
  to start below the fold
- Make day of the week headings white accordingly and increase padding

rileysw committed:
added more detailed descriptions to meal events

taesungh committed:
Fix font size and weight in schedule countdown

taesungh committed:
Fix past Announcement card style

samderanova commented on src/views/schedule/components/EventSpacerCard.tsx:
```diff
@@ -0,0 +1,43 @@
+	const minHeight = 80 * (duration + 1);
+
+	return (
+		<Card
+			className={"text-end mb-4 " + (over && styles.pastEventCard)}
+			bg="dark"
+			text="white"
+			style={{ minHeight }}
+		>
+			<Card.Body className="d-flex flex-column justify-content-center">
+				<Card.Title as="h4">{title}</Card.Title>
+				<Card.Text>{description}</Card.Text>
+				<footer className="text-white-50 text-end">{durationText}</footer>
+			</Card.Body>
+		</Card>
```
It might just be me, but all of these are aligned to the right instead of the center:
![image](https://user-images.githubusercontent.com/28026520/216682306-142c28c8-d7ba-47d1-9edc-15a8837e90a9.png)

taesungh replied:
That's part of the style we used since HackUCI 2021. Open to other suggestions though.

Pegysus reviewed (approved):
Looks good, thanks for working on this and fixing it up!

taesungh reviewed (approved):
Thanks for the help everyone. Some things with the countdown could be a little cleaner but this works for now. We may consider reintroducing scrollspy as well.
